nat-lab/tests/utils/batching.py

```diff
@@ -0,0 +1,161 @@
+# There's a dillemma in how to best capture the traffic. Capturing on the node itself
+# provides insights into the packets themselves coming to and from the tunnel device,
+# however capturing locally doesn't mean the packet actually left the machine.
+# An alternative is to capture on gateways which guarantees the packet arrival
+# however there's no way to easily pry open the packets and observe their insides.
+import asyncio
+import math
+import os
+import pytest
+import subprocess
+import typing
+from scapy.all import PcapReader  # type: ignore
+from scapy.layers.inet import IP, ICMP  # type: ignore
+from typing import Any, Callable, List, Tuple
+
+
+class Histogram:
+    def __init__(self) -> None:
+        self._values: typing.List[float] = []
+
+    def add_value(self, value: float) -> None:
+        self._values.append(value)
+
+    def get(self, buckets: int, bucket_size: int) -> List[int]:
+        if len(self._values) == 0:
+            return []
+
+        sorted_values = sorted(self._values)
+        max_val = sorted_values[-1]
+
+        if max_val >= buckets * bucket_size:
+            raise ValueError(
+                f"Histogram doesn't fit the data({max_val}). The max value is {max_val} but the histogram has {buckets} buckets with a width of"
+                f" {bucket_size}, the maximum value it can fit is {buckets * bucket_size}"
+            )
+
+        hs = [0] * buckets
+
+        for v in sorted_values:
+            bucket_index = int(v / bucket_size)
+            hs[bucket_index] += 1
+
+        return hs
+
+    def __repr__(self) -> str:
+        return f"Histogram of {len(self._values)} values"
+
+
+async def capture_histograms(
+    container_name: str, duration_s: int, filters: list["PacketFilter"]
+) -> List[Tuple[str, List[int]]]:
+    cmd_rm = f"docker exec --privileged {container_name} rm /home/capture.pcap"
+    os.system(cmd_rm)
+
+    iface = "any"
+    capture_path = "/home/capture.pcap"
+
+    cmd = f"docker exec -d --privileged {container_name} tcpdump -i {iface} -U -w {capture_path}"
+    res = os.system(cmd)
+    if res != 0:
+        raise RuntimeError(f"Failed to launch tcpdump on {container_name}")
+
+    await asyncio.sleep(duration_s)
+
+    local_path = f"{container_name}.pcap"
+    subprocess.run([
+        "docker",
+        "cp",
+        container_name + ":" + "/home/capture.pcap",
+        local_path,
+    ])
+
+    histograms = []
+    for f in filters:
+        # add extra time for histogram as tcpdump was not killed immediately
+        hs = _generate_histogram(local_path, duration_s + 1, 1, f)
+        histograms.append((f._name, hs))
+
+    os.unlink(local_path)
+
+    cmd_rm = f"docker exec --privileged {container_name} pkill tcpdump"
+    os.system(cmd_rm)
+
+    return histograms
+
+
+# Render ASCII histogram drawing for visual inspection
+def draw_histogram(name: str, data: List[int], max_height=None):
```

Comment:
Why isn't it a method of `Histogram`?

Reply:
This becomes irrelevant since I removed the `Histogram` class entirely and renamed `get` into `generate_histogram -> list[int]`